Customer search: Fill customer fields on Order screen

iamgabrielma opened the pull request:
Part of: #7741
Closes: https://github.com/woocommerce/woocommerce-ios/issues/7941

### Description
This PR deals with: 
- [x] Shows an empty view when we start a new keyword search, for parity with Android.
- [x] Fill the customer details into the Order details fields once a Customer is selected.
- [x] Adds a new view when no Customer search results are found.
- [x] Ignores searches for non-registered customers.

### Changes

- The `OrderCustomerListView` receives a completion handler with a `Customer` object when a merchant taps in the search button. Then we use this data to fill the `EditOrderAddressForm` viewModel.
- Removes `customerSearchTapped()`, as we're not using it anymore.
- When mapping `SearchResults` to `Customer` objects, we skip mapping those with `id = 0` (unregistered), avoiding requests to `/wc/v3/customers/0` that will always fail.

### Testing instructions

1. Go to Orders > + > Create new Order > + Add Customer details > Tap on the magnifying glass icon.
2. Search for a Customer name, and tap on it.

| | | 
|---|---|
| <img width=400 src="https://user-images.githubusercontent.com/3812076/192192292-26a940ff-353f-4c1b-99ba-54d443c0a91d.png"> | <img width=400 src="https://user-images.githubusercontent.com/3812076/198179154-8f594169-dfef-41e2-ab3d-34be43a0c462.png"> |


3. Confirm that Customer fields are filled (will fill both Shipping/Billing if these are different as well).
4. When searching for non-matching keywords, confirm that the empty search results show up.

| | | 
|---|---|
| <img width=400 src="https://user-images.githubusercontent.com/3812076/198179552-a529b898-70ae-4555-bed4-ed59c487270d.png"> | <img width=400 src="https://user-images.githubusercontent.com/3812076/198178774-fde668c6-d88c-4eb9-ba4a-0b269878dabf.png"> |


---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

iamgabrielma committed:
fill Customer fields on Order screen

iamgabrielma committed:
Merge branch 'trunk' into issue/7741-fill-order-data-from-customer-search

iamgabrielma committed:
Add shipping fields to EditAddressOrderForm

iamgabrielma committed:
Add CustomerSearchUICommand unit tests

iamgabrielma committed:
Hound: trailing_newline

iamgabrielma committed:
Internal comment cleanup

iamgabrielma committed:
Skip searchCustomer request for customerID = 0

iamgabrielma committed:
Remove unnecessary method

iamgabrielma committed:
Add empty view for empty search results

iamgabrielma committed:
Set empty state for new keyword searches

iamgabrielma commented on WooCommerce/WooCommerceTests/ViewRelated/Search/Customer/CustomerSearchUICommandTests.swift:
```diff
@@ -0,0 +1,46 @@
+
+        // When
+        let predicate = command.searchResultsPredicate(keyword: "some")
+        let expectedQuery = "siteID == 123 AND ANY searchResults.keyword == \"some\""
```
Double check if this actually filters down multiple sites.

iamgabrielma replied:
Confirmed 👍 , for 2 customers with the same name and customer ID, in two different sites, only one is shown.

iamgabrielma commented on WooCommerce/WooCommerceTests/ViewRelated/Search/Customer/CustomerSearchUICommandTests.swift:
```diff
@@ -0,0 +1,46 @@
+        XCTAssertEqual(cellViewModel.title, "\(customer.firstName ?? "") \(customer.lastName ?? "")")
+        XCTAssertEqual(cellViewModel.subtitle, String(customer.email))
+    }
+}
```
Let's try to test if `createResultsController()` uses the correct Predicate.

iamgabrielma committed:
Extract fillCustomerFields method to viewModel

iamgabrielma committed:
Add Unit Test for fillCustomerFields()

iamgabrielma committed:
Display shipping details form if addresses differ

iamgabrielma commented:
Thanks for your time yesterday to go through the review/pairing @joshheald, I made the following changes:
- Extracted the `fillCustomerFields()` method to the viewModel: https://github.com/woocommerce/woocommerce-ios/pull/7885/commits/9cf838809632b5a1b4e785a10e8993fe4021d23b
- Added Unit Tests to check if fields are correctly being filled: https://github.com/woocommerce/woocommerce-ios/pull/7885/commits/ed043176432e6aebb2605f173c0fd09cbb72056b
- Make the shipping details form display automatically if addresses differ: https://github.com/woocommerce/woocommerce-ios/pull/7885/commits/5dffa9738ce7182ab10a65f5bacc772e06709fe7

joshheald commented on WooCommerce/Classes/ViewRelated/Orders/Order Details/Address Edit/AddressFormViewModelProtocol.swift:
```diff
@@ -401,6 +401,37 @@ open class AddressFormViewModel: ObservableObject {
+
+        fields.firstName = customer.firstName ?? ""
+        fields.lastName = customer.lastName ?? ""
+        fields.email = customer.email
```
```suggestion
        fields = populate(fields: fields, with: customer.billing)
        secondaryFields = populate(fields: secondaryFields, with: customer.shipping)
    }

    private func populate(fields: AddressFormFields, with address: Address?) -> AddressFormFields {
        var fields = fields

        fields.firstName = address.firstName ?? ""
        fields.lastName = address.lastName ?? ""
        fields.email = address.email
```

Instead of the name fields on customer, I think we should use `address.firstName` and `address.lastName`. Testing Android, with my always-strange test data with different names all over the place, shows that they use these fields to update the data.

This makes sense, because the shipping address may be to a different person, e.g. if it's a gift.

There's `address.email`, which could differ from `customer.email`. Android seems to use `billingAddress.email` there. The email for the secondary fields isn't shown on either platform... but I'm not sure whether it's used in any requests.

It's worth checking that: what happens if we save a different email address in each place? Does it get added to the order twice? Is that what Android does? Is it a bug, one way or the other?

It's also just a nicer function... "Populate address fields with address"

iamgabrielma replied:
That makes sense. I made the changes here: 121ba8a10f18fa7b0f78f998df81c82d6b186544

Something that peaks my interest with this change is that the email property is now an optional, but it can't actually be optional because is a mandatory field for shippable orders. This seems to happen because when we declare both `fields` and `secondaryFields` in the protocol, both inherit from `AddressFormFields`, however these have slightly different parameters in the API (for example shipping details does not have an email field)

> what happens if we save a different email address in each place? Does it get added to the order twice? Is that what Android does? Is it a bug, one way or the other?

This seems a bug from our end when declaring the model, as with the `email` example above, we're reusing the same `AddressFormFields` model for both shipping and billing, but these are not exactly the same in the API or WC core (it would seem that the API doc is outdated as well, as phone does not appear as property for shipping but in WC core it does). I've opened an issue here: https://github.com/woocommerce/woocommerce-ios/issues/7993 as most likely this can be improved.

joshheald commented on WooCommerce/Classes/ViewRelated/Orders/Order Details/Address Edit/EditOrderAddressForm.swift:
```diff
@@ -172,9 +172,8 @@ struct EditOrderAddressForm<ViewModel: AddressFormViewModelProtocol>: View {
-                print("3 - Customer Callback. Fill Order data with Customer details")
-                print("4 - Customer ID: \(customer.customerID) - Name: \(customer.firstName ?? ""))")
+                viewModel.customerSelectedFromSearch(customer: customer)
+                showingCustomerSearch = false
```
This can move to the view model too if you like? Can wait for a future PR

joshheald commented on WooCommerce/WooCommerceTests/ViewRelated/Orders/Order Details/Addresses/EditOrderAddressFormViewModelTests.swift:
```diff
@@ -716,6 +716,76 @@ final class EditOrderAddressFormViewModelTests: XCTestCase {
+        viewModel.customerSelectedFromSearch(customer: customer)
+
+        // Then
+        XCTAssertEqual(viewModel.fields.email, customer.email)
```
Don't forget about `assertEqual`: the nicer diffs when stuff is wrong are worth it, even though it's annoying having to swap around the `actual` and `expected` when you swap from `XCTAssertEqual`

joshheald commented on WooCommerce/WooCommerceTests/ViewRelated/Orders/Order Details/Addresses/EditOrderAddressFormViewModelTests.swift:
```diff
@@ -716,6 +716,76 @@ final class EditOrderAddressFormViewModelTests: XCTestCase {
+        XCTAssertEqual(viewModel.fields.phone, customer.shipping?.phone)
+    }
+
+    func test_OrderAddressForm_shows_different_address_form_fields_when_addresses_differ_and_customerSelectedFromSearch() {
+        // Given
+        let viewModel = EditOrderAddressFormViewModel(order: Order.fake(), type: .billing)
+        let billing = sampleAddressWithEmptyNullableFields()
+        let shipping = Address.fake().copy(address1: "123 different fake street")
+        let customer = Customer.fake().copy(billing: billing, shipping: shipping)
+
+        // When
+        viewModel.customerSelectedFromSearch(customer: customer)
+
+        // Then
+        XCTAssertTrue(viewModel.showDifferentAddressForm)
+    }
```
👏👏👏 So good to be able to test this kind of thing with confidence

joshheald commented on WooCommerce/WooCommerceTests/ViewRelated/Search/Customer/CustomerSearchUICommandTests.swift:
```diff
@@ -0,0 +1,46 @@
+
+    override func tearDown() {
+        super.tearDown()
+    }
```
You can delete these

iamgabrielma replied:
Done! 7b58531d1272643fddc848ed904c820ef2b2b7e4

joshheald commented on Yosemite/YosemiteTests/Stores/CustomerStoreTests.swift:
```diff
@@ -184,4 +184,23 @@ final class CustomerStoreTests: XCTestCase {
+        // Then
+        let customers = try response.get()
+        XCTAssertEqual(customers.count, 0)
+    }
```
Nice test! This is business logic which is well worth checking.

However... it doesn't really test what it looks like. I had a go at making it fail (by removing the `if result.userID == 0` check in the store) and it still passed.... When you're testing a negative, it's especially worthwhile making sure that the test fails when you expect. Think Red-Green-Refactor, even when you're not actually doing TDD.

The reason is that we don't have anyone in `wc-analytics-customer.json` with `user_id: 0`, you'll need to add another entry to that file to get the test to work.

We can improve it further though... here's another way to do it, which more directly checks that the request is not sent, without needing to see what the customer response is:

(N.B. This still needs an analytics customer with `user_id: 0` in the json file)

```suggestion
    func test_searchCustomers_returns_no_customers_when_customer_is_not_registered() throws {
        // Given
        network.simulateResponse(requestUrlSuffix: "customers", filename: "wc-analytics-customers")

        // When
        waitFor { promise in
            let action = CustomerAction.searchCustomers(siteID: self.dummySiteID, keyword: self.dummyKeyword) { result in
                promise(())
            }
            self.dispatcher.dispatch(action)
        }

        // Then
        let requests = network.requestsForResponseData.compactMap { $0 as? JetpackRequest }
        XCTAssertFalse(requests.contains(where: { request in
            request.path == "customers/0"
        }))
    }
```

iamgabrielma replied:
Oh that's nice! Made the changes on 7b58531d1272643fddc848ed904c820ef2b2b7e4

joshheald reviewed (approved):
Pretty much there on this, good work.

I noticed another difference from Android (the fields used for name and email) and a problem with one of your tests. A couple of other small comments too.

iamgabrielma committed:
Add/improve test case for unregistered customer

iamgabrielma committed:
Populate fields with Address details, not Customer

We’ll be populating the fields using the Address data rather than Customer data for these 3 fields, for consistency with Android. There’s some inconsistency with the API though, reported here: https://github.com/woocommerce/woocommerce-ios/issues/7993

iamgabrielma committed:
Update WCAnalyticsCustomerMapperTests

iamgabrielma committed:
Update EdirOrderAddressFormViewModel tests
